### ISSUE TITLE
Background periodic publisher doesn't recover from a network exception

### ISSUE THREAD

**Issue by coderfromhere** [mtth/tracing#8]
If a tracing server is temporarily down, the background thread still is expected to survive after `flushSpans` throws `ConnectionFailure`:
```
HttpExceptionRequest Request {
  host                 = "localhost"
  port                 = 9411
  secure               = False
  requestHeaders       = [("content-type","application/json")]
  path                 = "/api/v2/spans"
  queryString          = ""
  method               = "POST"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (ConnectionFailure Network.Socket.connect: <socket: 54>: does not exist (Connection refused))
```

**Comment by mtth** [comment]
Agreed - the current behavior is not great. The background thread should fail the whole process on error ([relevant read](https://www.fpcomplete.com/blog/2017/08/exiting-haskell-process/)) or continue publishing.

In the meantime here are a couple suggestions to work around this:

1. Call `publish` manually with adequate error handling.
2. (Untested) Specify a custom request manager which [retries on a subset of exceptions](https://hackage.haskell.org/package/http-client-0.7.6/docs/Network-HTTP-Client.html#v:managerRetryableException).

**Comment by avanov** [comment]
> The background thread should fail the whole process on error (relevant read) or continue publishing.

Right, the only viable option in case of backend daemons is to carry on with (or without) delayed retrying to send the same payload again, as failing the entire process is hardly desirable. How about performing another `forkIO` with a retrying-only closure upon receiving a network exception? The number of retries could then be configured similarly to `settingsPublishPeriod`.
